### ISSUE TITLE
Confusing Error "Connection Error While Uploading Answers" - Due to known OSM read only mode

### ISSUE THREAD

**Issue by peternewman** [streetcomplete/StreetComplete#6045]
Due to https://community.openstreetmap.org/t/openstreetmap-org-currently-offline-operations-team-are-working-to-restore-15-december-2024/122930 the OSM database is currently in read only mode.

Download currently works fine, but upload fails with "Connection Error While Uploading Answers". If you look in the log, you get a more informative message that it's a 503 with "Database offline for maintenance".

**How to Reproduce**
Have an outage such as https://community.openstreetmap.org/t/openstreetmap-org-currently-offline-operations-team-are-working-to-restore-15-december-2024/122930

**Expected Behavior**
While I appreciate it wouldn't be translated, it might be nice to show the error from the server in a more direct way than having to dig in the logs.

I also wonder if there should be a pair of error messages, with one trying to match an inability to reach the API (e.g. due to being on a private network, or stuck behind a captive portal) and another when there is a problem with the API itself (e.g. 5xx errors) so they could ideally be a bit more descriptive.

Maybe even try accessing westnordost's website too in the event of a failure to try and narrow down a likely cause (given SC is in the fairly unique position of having two independently hosted APIs it can hit/test).

**Versions affected**
v59.3

**Comment by matkoniecz** [comment]
>  While I appreciate it wouldn't be translated, it might be nice to show the error from the server in a more direct way than having to dig in the logs.

Note that error from the server is actually more misleading and confusing as database is offline not really for maintenance but awaits (for multiple days!) replacement of broken router from ISP (on ISP side, they plugged two independent lines into the same device for which they had no working spares).

**Comment by taxmeifyoucan** [comment]
+1 It's confusing to display a general connection error when the issue is known to be on the server, I was checking my network settings and VPN before figuring it out in app logs. 

> on ISP side, they plugged two independent lines into the same device for which they had no working spares
Logs show server returning 503 so backend is offline but it would be hilarious to have this specific error :)

**Comment by westnordost** [comment]
I think the current error is just fine. For the user it doesn't matter whether the server responds with an "Error, but it's my fault!", if the server responds with garbage data or if the server is not reachable at all due to some network error. From the perspective of the user, it doesn't work because for one reason or another, the server cannot be connected to. And in all of these cases, what should the user should do is to try again later. I believe this is what is being shown to the user, right? That he should try again later?

**Comment by matkoniecz** [comment]
@taxmeifyoucan suggests to distinguish "your internet is broken altogether, maybe check whether you paid for mobile internet etc" from "OSM servers cannot be reached, try later"

**Comment by mnalis** [comment]
> From the perspective of the user, it doesn't work because for one reason or another, the server cannot be connected to

Well, I might agree as far as "if the server returns the error, it doesn't matter to user exactly what the error is" (although, as an techie, I'd personally still like to know!)

But, I **disagree that it doesn't matter whether it is network error or server error**. Although user cannot do anything about server errors, even relatively many non-technical users can try to do something about connection errors (e.g. disable wifi and enable mobile data or vice versa; disable their VPN etc).

---

In general, I think **dumbification of error messages is not helpful**. Not for newbie experts (who are eventually going to try to ask someone more knowledgeable, but would then be unable to provide details, so nobody will be able to help them) and especially not for more advanced users.

Even if one were to still insist on showing dumbified error message in order not to frighten the users or whatever, they should then provide button _"Technical information details"_ which shows actual low-level error cause (e.g. Socket error number, verbatim API server response etc)

**Comment by matkoniecz** [comment]
Not sure whether I want to go as far as to declare that such debug should be definitely always available.

**Comment by westnordost** [comment]
Ok, I'd be fine to display two different error messages for the user, distinguishing between "cannot reach, try again later" and "server problem, try again later".

**Comment by matkoniecz** [comment]
It seems that

```
    private suspend fun getMapDataOrNull(query: String): MapData? = wrapApiClientExceptions {
        try {
            val response = httpClient.get(baseUrl + query) { expectSuccess = true }
            val source = response.bodyAsChannel().asSource().buffered()
            return parser.parseMapData(source, emptySet())
        } catch (e: ClientRequestException) {
            when (e.response.status) {
                HttpStatusCode.Gone, HttpStatusCode.NotFound -> return null
                else -> throw e
            }
        }
    }
```

and

```
    suspend fun uploadChanges(
        changesetId: Long,
        changes: MapDataChanges,
        ignoreRelationTypes: Set<String?> = emptySet()
    ): MapDataUpdates = wrapApiClientExceptions {
        try {
            val response = httpClient.post(baseUrl + "changeset/$changesetId/upload") {
                userAccessTokenSource.accessToken?.let { bearerAuth(it) }
                setBody(serializer.serialize(changes, changesetId))
                expectSuccess = true
            }
            val source = response.bodyAsChannel().asSource().buffered()
            val updates = parser.parseElementUpdates(source)
            val changedElements = changes.creations + changes.modifications + changes.deletions
            return createMapDataUpdates(changedElements, updates, ignoreRelationTypes)
        } catch (e: ClientRequestException) {
            when (e.response.status) {
                // current element version is outdated or current changeset has been closed already
                HttpStatusCode.Conflict,
                // an element referred to by another element does not exist (anymore) or was redacted
                HttpStatusCode.PreconditionFailed,
                // some elements do not exist anymore as it was deleted
                HttpStatusCode.Gone,
                // some elements do not exist and never existed
                HttpStatusCode.NotFound -> {
                    throw ConflictException(e.message, e)
                }
                HttpStatusCode.PayloadTooLarge -> {
                    throw ChangesetTooLargeException(e.message, e)
                }
                else -> throw e
            }
        }
    }
```

both in https://github.com/streetcomplete/StreetComplete/blob/2c4303539600c176c8adcb7d1b831edad128496d/app/src/commonMain/kotlin/de/westnordost/streetcomplete/data/osm/mapdata/MapDataApiClient.kt would need to start throwing different exception based on response

the tricky part is that testing it would be a bit annoying (mocking is possible - but which exact error/code is given by real API?)

Luckily we have

> If you look in the log, you get a more informative message that it's a 503 with "Database offline for maintenance".

503 is for "Service Unavailable" so all seems to make sense

**Comment by westnordost** [comment]
It's not that tricky. See for example `StatisticsApiClientTest.kt` for how to mock a `HttpClient`. It's very easy, as Ktor already comes with the necessary utilities for that.

**Comment by matkoniecz** [comment]
I was rather worried about mocked response not matching actual server behaviour.

**Comment by mnalis** [comment]
> Luckily we have
>>    If you look in the log, you get a more informative message that it's a 503 with "Database offline for maintenance".
> 503 is for "Service Unavailable" so all seems to make sense

Exactly... And we have other examples of errors around if you look for other `report_error` uses:
https://github.com/openstreetmap/openstreetmap-website/blob/5a07336e50aa6efa4d1d425eaec0a5451dbde102/app/controllers/application_controller.rb#L108

it might be `:forbidden` (403) instead of `:service_unavailable` (503) as well. 

But it probably makes sense to handle **all non-OK** (4xx, 5xx) API responses in the same way.

I would suggest always logging (in app log) both the error code and exact reason (content body) returned.

As for to what will be displayed to the user, I personally would prefer if show literal error body after introduction text (e.g. _"Unable to upload because server returned error: **Database offline for maintenance**"_), but I'm OK if just a fixed string (e.g. _"Unable to because server returned error"_) is shown to the user -- if people _really_ feel it is more user friendly[^1] (as long as exact error is available in the app log).


[^1]: I personally feel dumbification of error messages is **less** user friendly; as when they report the problem we will probably want to know why the error happened, and then instead of just sharing screenshot of the error (or transcribing it), they'll have to find how to fetch error log from the app and share that, which is more complex, might require filtering and may not work after all (e.g. https://github.com/streetcomplete/StreetComplete/issues/5561)
